### PR TITLE
Port rocket/Frontend to chisel3

### PR DESCRIPTION
**Pull request by CircuitCoder**
Type of change: other enhancement

Impact: no functional change

Development Phase: implementation

### CODE REVIEW

**Comment by sequencer** [comment]
rebase plz

File: src/main/scala/rocket/Frontend.scala [hunk 1]
```diff
@@ -148,17 +148,18 @@ class FrontendModule(outer: Frontend) extends LazyModuleImp(outer)
...
-  when(io.ptw.req.fire && recent_progress) { recent_progress_counter := recent_progress_counter - 1 }
+  val recent_progress = recent_progress_counter > 0.U
+  when(io.ptw.req.fire && recent_progress) { recent_progress_counter := recent_progress_counter - 1.U }
   when(io.cpu.progress) { recent_progress_counter := recent_progress_counter_init }
 
   val s2_kill_speculative_tlb_refill = s2_speculative && !recent_progress
 
   io.ptw <> tlb.io.ptw
   tlb.io.req.valid := s1_valid && !s2_replay
+  tlb.io.req.bits.cmd := M_XRD // Frontend only reads
```
Comment (sequencer) [inline src/main/scala/rocket/Frontend.scala:159]: This must be a subtle bug...
Thanks for figuring this out

File: src/main/scala/rocket/Frontend.scala [hunk 2]
```diff
@@ -318,55 +325,55 @@ class FrontendModule(outer: Frontend) extends LazyModuleImp(outer)
...
   }
   when (io.cpu.req.valid) {
-    gpa_valid := false
+    gpa_valid := false.B
   }
   io.cpu.gpa.valid := gpa_valid
   io.cpu.gpa.bits := gpa
 
   // performance events
-  io.cpu.perf := icache.io.perf
```
Comment (sequencer) [inline src/main/scala/rocket/Frontend.scala:356]: why

Comment (CircuitCoder) [reply src/main/scala/rocket/Frontend.scala:356]: perf on the CPU side have an additional field `tlbMiss` (See below)
This is essentially a partial assignment.

**Pull request merged**